p2p/transport.go

```diff
@@ -0,0 +1,488 @@
+				_ = c.Close()
+				return
+			}
+		}(c)
+	}
+}
+
+func (mt *MultiplexTransport) cleanup(c net.Conn) error {
+	mt.conns.Remove(c)
+
+	return c.Close()
+}
+
+func (mt *MultiplexTransport) filterConn(c net.Conn) (err error) {
+	defer func() {
+		if err != nil {
+			_ = c.Close()
+		}
+	}()
+
+	// Reject if connection is already present.
+	if mt.conns.Has(c) {
+		return ErrRejected{conn: c, isDuplicate: true}
+	}
+
+	// Resolve ips for incoming conn.
+	ips, err := resolveIPs(mt.resolver, c)
+	if err != nil {
+		return err
+	}
+
+	errc := make(chan error, len(mt.connFilters))
+
+	for _, f := range mt.connFilters {
+		go func(f ConnFilterFunc, c net.Conn, ips []net.IP, errc chan<- error) {
+			errc <- f(mt.conns, c, ips)
+		}(f, c, ips, errc)
+	}
+
+	for i := 0; i < cap(errc); i++ {
+		select {
+		case err := <-errc:
+			if err != nil {
+				return ErrRejected{conn: c, err: err, isFiltered: true}
+			}
+		case <-time.After(mt.filterTimeout):
+			return ErrFilterTimeout{}
+		}
+
+	}
+
+	mt.conns.Set(c, ips)
+
+	return nil
+}
+
+func (mt *MultiplexTransport) upgrade(
+	c net.Conn,
+) (sc *conn.SecretConnection, ni NodeInfo, err error) {
+	defer func() {
+		if err != nil {
+			_ = mt.cleanup(c)
+		}
+	}()
+
+	sc, err = secretConn(c, mt.handshakeTimeout, mt.nodeKey.PrivKey)
+	if err != nil {
+		return nil, NodeInfo{}, ErrRejected{
+			conn:          c,
+			err:           fmt.Errorf("secrect conn failed: %v", err),
+			isAuthFailure: true,
+		}
+	}
+
+	ni, err = handshake(sc, mt.handshakeTimeout, mt.nodeInfo)
+	if err != nil {
+		return nil, NodeInfo{}, ErrRejected{
+			conn:          c,
+			err:           fmt.Errorf("handshake failed: %v", err),
+			isAuthFailure: true,
+		}
+	}
+
+	if err := ni.Validate(); err != nil {
+		return nil, NodeInfo{}, ErrRejected{
+			conn:              c,
+			err:               err,
+			isNodeInfoInvalid: true,
+		}
+	}
+
+	// Ensure connection key matches self reported key.
+	if connID := PubKeyToID(sc.RemotePubKey()); connID != ni.ID {
+		return nil, NodeInfo{}, ErrRejected{
+			conn: c,
+			id:   connID,
+			err: fmt.Errorf(
+				"conn.ID (%v) NodeInfo.ID (%v) missmatch",
+				connID,
+				ni.ID,
+			),
+			isAuthFailure: true,
+		}
+	}
+
+	// Reject self.
+	if mt.nodeInfo.ID == ni.ID {
+		return nil, NodeInfo{}, ErrRejected{conn: c, id: ni.ID, isSelf: true}
+	}
+
+	if err := mt.nodeInfo.CompatibleWith(ni); err != nil {
+		return nil, NodeInfo{}, ErrRejected{
+			conn:           c,
+			err:            err,
+			id:             ni.ID,
+			isIncompatible: true,
+		}
+	}
+
+	return sc, ni, nil
+}
+
+func (mt *MultiplexTransport) wrapPeer(
+	c net.Conn,
+	ni NodeInfo,
+	cfg peerConfig,
+) Peer {
+	p := newPeer(
+		peerConn{
+			conn:       c,
+			config:     &mt.p2pConfig,
+			outbound:   cfg.outbound,
+			persistent: cfg.persistent,
+		},
+		mt.mConfig,
+		ni,
+		cfg.reactorsByCh,
+		cfg.chDescs,
+		cfg.onPeerError,
+	)
+
+	// Wait for Peer to Stop so we can cleanup.
+	go func(c net.Conn) {
+		select {
```

Comment:
you don't need select if there's just one blocking case.

Reply:
![facepalm](https://media.giphy.com/media/3og0INyCmHlNylks9O/giphy.gif)